### PR TITLE
Refactoring of QueryGraph

### PR DESCRIPTION
**Pull request by systay**

### CODE REVIEW

**Comment by pontusmelke** [comment]
ha?

**Comment by systay** [comment]
@pontusmelke `ha?`?

File: community/cypher/ir-3.4/src/main/scala/org/neo4j/cypher/internal/ir/v3_4/QueryGraph.scala [hunk 1]
```diff
@@ -25,6 +25,12 @@ import org.neo4j.cypher.internal.v3_4.expressions.{Expression, LabelName, Proper
 
 import scala.collection.{GenTraversableOnce, mutable}
 
+/*
+This is one of the core classes used during query planning. It represents the declarative query,
+it contains no more information that the AST, but it contains data in a format that is easier
+to consume by the planner. If you want to trace this back to the original query - one QueryGraph
+represents all the MATCH, OPTIONAL MATCHes, and update clauses between two WITHs.
+ */
```
Comment (sherfert) [inline community/cypher/ir-3.4/src/main/scala/org/neo4j/cypher/internal/ir/v3_4/QueryGraph.scala:34]: Having such comments really helps understand our (planner) code faster. Nice.

File: community/cypher/ir-3.4/src/main/scala/org/neo4j/cypher/internal/ir/v3_4/QueryGraph.scala [hunk 2]
```diff
@@ -218,33 +177,32 @@ case class QueryGraph(patternRelationships: Set[PatternRelationship] = Set.empty
...
-  def coveredIds: Set[IdName] = {
-    val patternIds = QueryGraph.coveredIdsForPatterns(patternNodes, patternRelationships)
-    patternIds ++ argumentIds ++ selections.predicates.flatMap(_.dependencies)
-  }
+  /**
+    * Variables are bound after matching this QG, but before optional
+    * matches and updates have been applied
+    */
+  def idsWithoutOptionalMatchesOrUpdates: Set[IdName] =
+    QueryGraph.coveredIdsForPatterns(patternNodes, patternRelationships) ++ argumentIds
```
Comment (sherfert) [inline community/cypher/ir-3.4/src/main/scala/org/neo4j/cypher/internal/ir/v3_4/QueryGraph.scala:195]: Just to understand better, why was `selections.predicates.flatMap(_.dependencies)` removed?

**Review by sherfert — APPROVED** [review]

File: community/cypher/ir-3.4/src/main/scala/org/neo4j/cypher/internal/ir/v3_4/QueryGraph.scala (hunk 2 above)
Comment (systay) [reply community/cypher/ir-3.4/src/main/scala/org/neo4j/cypher/internal/ir/v3_4/QueryGraph.scala:195]: The expression that was removed was actually not listing variables exposed, but instead which dependencies the predicates have. I changed https://github.com/neo4j/neo4j/pull/10533/files#diff-0ef12799d6f97a3d6751a5ced882c8b9R81 to use the dependencies explicitly. That is also why there are so many failures - we relied on this weirdness in a bunch of places. I'll add at least one more commit to this PR

**Comment by fickludd** [comment]
This is great, any added clarity around the mysterious `QueryGraph` is a leap in the right direction. (I'm still mystified, but slightly less so)

**Comment by pontusmelke** [comment]
@systay it was labeled with ha, but not any longer. Move along, nothing to see here.

**Pull request merged**